### PR TITLE
Fix Sigma list issues during "pending" mint/spend (ie. mempool)

### PR DESCRIPTION
**Pull request by riordant**
Merged with https://github.com/zcoinofficial/zcoin/pull/504 to fix conflicts, please merge that PR first.

### CODE REVIEW

File: src/hdmint/tracker.h [hunk 1]
```diff
@@ -52,7 +53,7 @@ class CHDMintTracker
     bool UnArchive(const uint256& hashPubcoin, bool isDeterministic);
     bool UpdateZerocoinEntry(const CSigmaEntry& zerocoin);
     bool UpdateState(const CMintMeta& meta);
-    void SetMetaNonDeterministic();
+    std::set<uint256> GetMempoolTxids();
```
Comment (thebevrishot) [inline src/hdmint/tracker.h]: does not make sense to define here as public function. I think you should move to `src/main.cpp` as static function or set it as private to allow only internal use.

**Comment by ultimaweapon** [comment]
Wait, if we merge this PR before #504 that mean we don't need #504 anymore due to all changes in that PR already merged with this PR. This does not look right to me.

**Comment by riordant** [comment]
That is why is said in the description to merge the other one first.

**Comment by ultimaweapon** [comment]
Sorry for misunderstood. I though `this` mean this PR.

File: qa/rpc-tests/hdmint_mempool_zap.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+
+# 1. start wallet
+# 2. mine blocks to activate sigma
+# 3. mint some coins
+# 4. generate blocks
+# 5. get listunspent transactions
+# 6. mint one more unconfirmed
+# 7. restart with `-zapwallettxes=1`
+# 8. check listunspentmints - should be as on step 5 (cause -zapwallettxes=1 clean mempool)
```
[the remark this answers is not in the document]
Comment (riordant) [reply qa/rpc-tests/hdmint_mempool_zap.py]: After clearing the mempool the mint will reappear in listunspentmints on next run.

File: src/hdmint/tracker.cpp [hunk 3]
```diff
@@ -49,19 +49,14 @@ bool CHDMintTracker::Archive(CMintMeta& meta)
...
-    //     //failed to read mint from DB, try reading deterministic
-    //     CHDMint dMint;
-    //     if (!walletdb.ReadHDMint(hashPubcoin, dMint))
-    //         return error("%s: could not find pubcoinhash %s in db", __func__, hashPubcoin.GetHex());
-    //     if (!walletdb.ArchiveDeterministicOrphan(dMint))
-    //         return error("%s: failed to archive deterministic ophaned mint", __func__);
-    // }
+   CWalletDB walletdb(strWalletFile);
+   CSigmaEntry zerocoin;
+    //failed to read mint from DB, try reading deterministic
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/hdmint/tracker.cpp]: Agreed

**Comment by riordant** [comment]
Sure, sorry for miscommunication

File: qa/rpc-tests/hdmint_mempool_zap.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thebevrishot) [reply qa/rpc-tests/hdmint_mempool_zap.py]: ok, thanks

**Comment by ultimaweapon** [comment]
This is hard to review due to it also include changes from other PR. Could you merge it with latest `master` so it will exclude changes from merged PR?

**Comment by riordant** [comment]
Sure will do.

**Comment by riordant** [comment]
@ultimaweapon merged. @a-bezrukov can you re-approve (No changes in code, only merge with `master` to remove the same code)

**Pull request merged**